### ISSUE TITLE
Bot "tournaments" and statistics

### ISSUE THREAD

**Issue by wrongu** [Rochester-NRT/AlphaGo#69]
DeepMind reports win percentages of various versions of their models against each other and against Pachi, Fuego, etc.

Are there existing GTP controllers to run "tournaments" and get statistics?

Waiting on #52

**Comment by jtauber** [comment]
That might be a useful thing for me to focus on next for my GTP library.

**Comment by wrongu** [comment]
Is your library going to include a controller? I thought you were just making the engine part

**Comment by jtauber** [comment]
@wrongu I plan it to be a general GTP library for **both** controllers and engines. It won't itself be a full-blown engine as that requires an actual go-playing component but it will be suitable for interfacing with go-playing components to make them GTP engines. The controller side is much easier to do standalone and the "playing two engines against one another" is the classic controller use case which I'll get working on today.

**Comment by jtauber** [comment]
Okay, I have a little GTP controller script that uses my GTP library to play GnuGo against itself. It should work for other GTP engines too so I'm going to try that next.

**Comment by jtauber** [comment]
The GTP controller script is available at https://github.com/jtauber/gtp/blob/master/gnugo_vs_gnugo.py

I plan to refactor out the reusable code and ship it with the library itself.

**Comment by robert-waite** [comment]
Just wanted to comment that I trained a SL policy network and wanted to run it against GnuGo. I used the gtp_wrapper.py provided in this project and then used the "ringmaster" program provided by GoMill. Seems like it provides much of the functionality that would be desired. 

The AlphaGo paper seemed to rely on running against other Go programs to evaluate and make decisions about where they wanted to go next... so seems very important for anyone wanting to tweak the training pipeline.

I did notice however that GnuGo would resign with the current responses from gtp_wrapper.py. GnuGo will make a pass... then RocAlphaGo will declare it illegal. I believe GnuGo then responds that RocAlphaGo is making an illegal move... and then GnuGo resigns... which means that from gomill's perspective... RocAlphaGo always beans GnuGo!

**Comment by wrongu** [comment]
@robert-waite something clearly went wrong here! At first-glance, [gomill appears to use the lowercase string "pass"](http://mjw.woodcraft.me.uk/gomill/doc/0.7.4/_modules/gomill/common.html#format_vertex) to denote passing moves. RocAlphago uses [jtauber's gtb library](https://github.com/jtauber/gtp) under the hood to process commands, which [should be able to handle "pass" as an input](https://github.com/jtauber/gtp/blob/master/gtp.py#L77).

If you're able to pinpoint the bug as something inside RocAlphaGo or the gtp library, please make a separate issue and/or a PR to fix it =)

**Comment by robert-waite** [comment]
Will play with it tonight and report back ;)

**Comment by apetresc** [comment]
Since nobody else has mentioned it: the folks at the [computer-go mailing list](http://computer-go.org/mailman/listinfo/computer-go) maintain a GTP-enabled go server specifically for playing bots against each other constantly, to get practice, data, and a stable rating. It's called CGOS: http://cgos.boardspace.net/

It's not very hard to plug into, and there's an active community of Go programmers (including, I believe, Aja Huang in the past) who play their bots there all the time. Would be a cool place to drop RocAlphaGo once it's ready.
